hq_superset/tests/test_utils.py

```diff
@@ -24,3 +25,90 @@ def test_doctests():
     import hq_superset.utils
     results = doctest.testmod(hq_superset.utils)
     assert results.failed == 0
+
+
+class TestDomainSyncUtil(SupersetTestCase):
+    PLATFORM_ROLE_NAMES = ["Gamma", "sql_lab", "dataset_editor"]
+
+    @patch.object(DomainSyncUtil, "_get_domain_access")
+    def test_gamma_role_assigned_for_edit_permissions(self, get_domain_access_mock):
+        security_manager = self.app.appbuilder.sm
+        self._ensure_platform_roles_exist(security_manager)
+
+        get_domain_access_mock.return_value = self._to_permissions_response(
+            can_write=True,
+            can_read=True,
+            roles=[],
+        )
+        additional_roles = DomainSyncUtil(security_manager)._get_additional_user_roles("test-domain")
+        assert len(additional_roles) == 1
+        assert additional_roles[0].name == "Gamma"
```

Comment:
should not the user always have the custom domain role?

Reply:
Not anymore, since we assign the Gamma role to "edit" users and will only create and assign a "read only" role to "read" users. 

I think I'm going to update the custom user domain role name such that it makes clear that it's a read-only role. So the name becomes `<domain>_user_<user_id>_read_only`

Reply:
https://github.com/dimagi/commcare-analytics/pull/59/commits/4d52e10f48937c255080e71371ec499e52c26bf1

Reply:
Oh you are just testing for the additional roles, sorry.

I was thinking about the domain role that is added to the user so they have access to just their domain?
This one, https://github.com/dimagi/commcare-analytics/blob/d8feb1d42c01b8b00ea8cc11bf81ae706c309d01/hq_superset/utils.py#L123

Reply:
> I think I'm going to update the custom user domain role name such that it makes clear that it's a read-only role. So the name becomes <domain>_user_<user_id>_read_only

Is that in line with the spec? We wanted to keep the naming consistent with roles. 
Additionally, this adds the extra overhead of changing the name to be without "read_only" if the permissions are updated, which is best avoided.

Reply:
The domain role that gives a user access to only their own domain is still given as usual and have nothing to do with these changes.

> Is that in line with the spec? We wanted to keep the naming consistent with roles.

Maybe not, but I thought it's probably a good idea to be able to see what permissions the role gives, since some users will have the `user_:user_id:_read_only` role and others not (I'm thinking in the scenario where someone's looking at a user's roles and want to know what access they have). If we give it a more generic name one would need to go and view the permissions individually to see what the role means.

Let me know if I need to change it back or if this new way is OK.

> Additionally, this adds the extra overhead of changing the name to be without "read_only" if the permissions are updated, which is best avoided.

It's just a one-liner, but yeah, that would need to update if we ever need to give more than read-only access that we decide to put into this role (as opposed to creating a different role).

Reply:
Hey @Charl1996 

> If we give it a more generic name one would need to go and view the permissions individually to see what the role means.

So, the roles like dataset_editor are different right? This one is just for read only and write access? So the role would either be read only or write?

When a user's permissions are upgraded from read only to write, do we change the name of the same role or a new one gets created & old one deleted?

> Let me know if I need to change it back or if this new way is OK.

I think its fine to keep it if you think its going to be useful on the UI. I am just unsure of the  benefit over the cost of the overhead it brings, though we can always revisit this later if needed.

Reply:
Actually,
Thinking more about this,
if the naming is updated to have `_read_only` in the role name, and if the users role is upgraded from view only to write, won't CCA end up having two roles when the syncing of role happens, since it will simply add a new role without the `_read_only` bit, leaving the old one around?

Reply:
@mkangia 

> When a user's permissions are upgraded from read only to write, do we change the name of the same role or a new one gets created & old one deleted?

We don't assign the role and assign the Gamma instead (leaving the old one lying around).

Now that you mention this, I'm beginning to think we should simplify this to follow the previous way where we simply assigned a user domain role which will contain all the permissions (read or write). The user will always have this role, but the permissions will only be different. 

In the case of an edit user we can simply assign all the gamma permissions to the role. This is as easy as
```
new_permissions = gamma_role.permissions
```
(why did I not think of this earlier?)

What do you think?

Reply:
I'd suggest to use the Gamma role itself.

> assign all the gamma permissions to the role

This seems like duplication of permissions.
Additionally, Superset/Flask architecture supports adding multiple roles, so seems good to leverage a role that already exists with necessary permissions instead of copying its permissions over.

Though, effectively both approaches have the same end result, so seems okay to do this if you feel that is better.

Its additionally about if there is any difference of effort in these two
1. assign gamma role permissions
2. remove gamma role from user if the user does not have edit access

Reply:
I'd like to add that for an "edit" user we assign the gamma role which also gives all the view permissions, so we don't assign the domain user role (the role that specifies which view/edit permissions the user has on a domain), meaning if a user was a "read only" user and is upgraded to the "edit" user there's a redundant view-only role in the DB that we'd need to clean up.

Reply:
See [this PR](https://github.com/dimagi/commcare-analytics/pull/71) for alternative approach.

Reply:
> if a user was a "read only" user and is upgraded to the "edit" user there's a redundant view-only role in the DB that we'd need to clean up

Oh yes, I missed that.